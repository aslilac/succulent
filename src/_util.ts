type IsAny<T> = unknown extends T ? (T extends {} ? T : never) : never;
type NotAny<T> = T extends IsAny<T> ? never : T;

export function assertType<T, X extends T>(x: NotAny<X>) {}

/**
 * This should always be an error (hence the ts-expect-error), and
 * is put here to make sure this is always true whenever assertType
 * is imported. Because assertType is a noop, the type error isn't
 * actually dangerous.
 */
// @ts-expect-error
assertType<unknown, any>({} as any);
