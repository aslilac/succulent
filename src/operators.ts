import { Schema, SchemaBase } from "./schema";

export function is<T>(x: unknown, schema: SchemaBase<T>): x is T {
	return Schema.check(schema, x);
}

type HasLength = { length: number };

export function hasLength(length: number) {
	return (x: HasLength) => x.length === length;
}

export function minLength(length: number) {
	return (x: HasLength) => x.length >= length;
}

export function maxLength(length: number) {
	return (x: HasLength) => x.length <= length;
}

export function nonEmpty(x: HasLength) {
	return x.length > 0;
}

type Filter<T> = (x: T) => boolean;

export function that<T>(schema: Schema<T>, ...filters: Array<Filter<T>>): Schema<T> {
	return new Schema(
		(x: unknown): x is T => is(x, schema) && filters.every((filter) => filter(x)),
	);
}

export function union<T extends readonly unknown[]>(
	...schemas: readonly [...Schema.WrapAll<T>]
): Schema<T[number]> {
	return new Schema(
		(t: unknown): t is T[number] => schemas.some((schema) => is(t, schema)),
		function* () {
			for (const schema of schemas) yield* Schema.from(schema);
		},
	);
}

export function or<X, Y>(x: SchemaBase<X>, y: SchemaBase<Y>): Schema<X | Y> {
	return new Schema((t: unknown): t is X | Y => is(t, x) || is(t, y));
}

export function and<X, Y>(x: SchemaBase<X>, y: SchemaBase<Y>): Schema<X & Y> {
	return new Schema((t: unknown): t is X & Y => is(t, x) && is(t, y));
}
