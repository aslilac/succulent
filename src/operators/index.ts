export * from "./is";
export * from "./iterables";
export * from "./logic";
