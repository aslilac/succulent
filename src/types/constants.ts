import { Schema } from "../schema";
import { $literal } from "./misc";

export const $boolean = new Schema((x: unknown): x is boolean => typeof x === "boolean", {
	displayName: "boolean",
});
export const $NaN = new Schema((x: unknown): x is typeof NaN => x !== x, {
	displayName: "NaN",
});

export const $false = $literal(false);
export const $true = $literal(true);
export const $undefined = $literal(undefined);
export const $null = $literal(null);
