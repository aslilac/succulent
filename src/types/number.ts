import { Schema } from "../schema.js";

/**
 * Checks if the value is a primitive `number`.
 * @example
 * ```ts
 * guard(1, $number); // ok!
 * guard(1n, $number); // will throw a `TypeError`
 * guard("1", $number); // will throw a `TypeError`
 * ```
 */
export const $number = new Schema(
	(x: unknown): x is number => typeof x === "number" && x === x,
	{ displayName: "number" },
);

/**
 * @example
 * ```ts
 * guard(1, $int); // ok!
 * guard(1.0, $int); // ok! (because `float` isn't really a type in JavaScript)
 * guard(1.1, $int); // will throw a `TypeError`
 * ```
 */
export const $int = new Schema((x: unknown): x is number => Number.isInteger(x), {
	displayName: "int",
});

/**
 * @example
 * ```ts
 * guard(1, $finite); // ok!
 * guard(2**128, $finite); // ok!
 * guard(1n, $finite); // will throw a `TypeError`, because `bigint` isn't a `number`
 * guard(Infinity, $finite); // will throw a `TypeError`, for more obvious reasons
 * guard(Number.NEGATIVE_INFINITY, $finite); // also throws a `TypeError`
 * ```
 */
export const $finite = new Schema((x: unknown): x is number => Number.isFinite(x), {
	displayName: "finite",
});

/**
 * @example
 * ```ts
 * guard(1n, $bigint); // ok!
 * guard(1, $bigint); // will throw a `TypeError`
 * ```
 */
export const $bigint = new Schema((x: unknown): x is bigint => typeof x === "bigint", {
	displayName: "bigint",
});
