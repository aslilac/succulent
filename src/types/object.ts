import { Schema, SchemaBase } from "../schema";

export const $anyobject = new Schema(
	(x: unknown): x is object => typeof x === "object" && x != null,
);

export function $object<T extends object>(template: {
	[K in keyof T]: SchemaBase<T[K]>;
}): Schema<T> {
	return new Schema(
		(x: unknown): x is T =>
			typeof x === "object" &&
			x != null &&
			// [
			// 	...Object.getOwnPropertyNames(template),
			// 	...Object.getOwnPropertySymbols(template),
			// ]
			Reflect.ownKeys(template).every((key) =>
				// @ts-expect-error - Can't quite get these types
				Schema.check(template[key], x[key]),
			),
	);
}
