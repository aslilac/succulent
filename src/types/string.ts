import { Schema } from "../schema";

export const $string = new Schema((x: unknown): x is string => typeof x === "string");
