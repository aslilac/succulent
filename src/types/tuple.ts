import { is } from "../operators";
import { Schema } from "../schema";

export function $tuple<T extends readonly unknown[]>(
	...schemas: readonly [...Schema.WrapAll<T>]
): Schema<T> {
	return (t: unknown): t is T =>
		Array.isArray(t) &&
		t.length === schemas.length &&
		schemas.every((schema, i) => is(t[i], schema));
}
