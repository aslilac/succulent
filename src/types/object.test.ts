import { assertType } from "../_util";
import { is, or, union } from "../operators";
import { $boolean, $undefined } from "./constants";
import { $number } from "./number";
import { $exact, $object } from "./object";
import { $string } from "./string";
import { $tuple } from "./tuple";

test("$object", () => {
	// The type signature of Symbol.for isn't quite correct, so we have to
	// assign the symbol we use here as a constant even though the whole point
	// of Symbol.for is to return the same symbol when called with the same
	// string multiple times.
	const key = Symbol("k");

	const example = $object({
		a: $boolean,
		b: $number,
		c: $string,
		d: $object({
			[key]: $tuple($number, $number, $number, $number),
			key: $tuple($number, $number, $number, $number),
		}),
	});

	const i = { a: true, b: 1, c: "hi", d: {} };
	const j = { ...i, d: { [key]: [0, 0, 0, 0] } };
	const k = { ...i, d: { key: [0, 0, 0, 0] } };

	const l = {
		...i,
		d: {
			[key]: [0, 0, 0, 0],
			key: [0, 0, 0, 0],
		},
	};

	expect(is(i, example)).toBe(false);
	expect(is(j, example)).toBe(false);
	expect(is(k, example)).toBe(false);
	expect(is(l, example)).toBe(true);

	type Example = { a: boolean; b: number; c: string; d: ExampleD };
	type ExampleD = { [key]: Numbers; key: Numbers };
	type Numbers = [number, number, number, number];
	function _(x: unknown) {
		if (is(x, example)) assertType<Example, typeof x>(x);
	}
});

test("$object with unwrapped literals", () => {
	expect(is({ hi: "hi" }, $object({ hi: "hi" }))).toBe(true);
	expect(is({ hi: "hi" }, $object({ hi: "hey" }))).toBe(false);
});

test("$object with optional keys", () => {
	// I think exactOptionalPropertyTypes would actually make this fail, but
	// I'm not sure what the solution would be if so.
	type Test = { hi: string; optional?: string };
	const schema = $object({ hi: $string, optional: union($string, $undefined) });

	expect(is({ hi: "hi" }, schema)).toBe(true);

	function _(x: unknown) {
		if (is(x, schema)) assertType<Test, typeof x>(x);
	}
});

test("Using $object to match an existing type", () => {
	type Friend = { name: string };

	// Validates that the object has a key of name with type string
	is<Friend>({}, $object({ name: $string }));

	// @ts-expect-error - Doesn't have name!
	is<Friend>({}, $object({ count: $number }));

	// Extra properties are fine
	is<Friend>({}, $object({ name: $string, count: $number }));

	// @ts-expect-error - string | undefined is not assignable to type string
	is<Friend>({}, $object({ name: or($undefined, $string) }));

	// This one works fine, because we're using Partial<Friend> instead of Friend
	is<Partial<Friend>>({}, $object({ name: or($undefined, $string) }));
});

test("$exact", () => {
	const schema = $exact({ a: $boolean, b: $boolean, c: $boolean });

	expect(is({ a: true, b: true }, schema)).toBe(false);
	expect(is({ a: true, b: true, c: true }, schema)).toBe(true);
	expect(is({ a: true, b: true, c: true, d: true }, schema)).toBe(false);
	expect(is({ b: true, c: true, d: true }, schema)).toBe(false);
});
