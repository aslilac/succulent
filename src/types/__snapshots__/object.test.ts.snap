// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$Exact 1`] = `
"Expected {|a: boolean, b: boolean, c: boolean|}, got [object Object]
  Expected property b to have type boolean
    Expected boolean, got undefined
  Expected property c to have type boolean
    Expected boolean, got undefined"
`;

exports[`$Exact 2`] = `
"Expected {|a: boolean, b: boolean, c: boolean|}, got [object Object]
  Unexpected property d
  Unexpected property e"
`;

exports[`$Exact 3`] = `
"Expected {|a: boolean, b: boolean, c: boolean|}, got [object Object]
  Unexpected property d"
`;

exports[`$interface 1`] = `
"Expected {a: boolean, b: number, c: string, d: {key: [number, number, number, number], [Symbol(k)]: [number, number, number, number]}}, got [object Object]
  Expected property d to have type {key: [number, number, number, number], [Symbol(k)]: [number, number, number, number]}
    Expected {key: [number, number, number, number], [Symbol(k)]: [number, number, number, number]}, got [object Object]
      Expected property key to have type [number, number, number, number]
        Expected [number, number, number, number], got undefined"
`;

exports[`$interface 2`] = `
"Expected {a: boolean, b: number, c: string, d: {key: [number, number, number, number], [Symbol(k)]: [number, number, number, number]}}, got [object Object]
  Expected property d to have type {key: [number, number, number, number], [Symbol(k)]: [number, number, number, number]}
    Expected {key: [number, number, number, number], [Symbol(k)]: [number, number, number, number]}, got [object Object]
      Expected property [Symbol(k)] to have type [number, number, number, number]
        Expected [number, number, number, number], got undefined"
`;
